--- fidget.nvim/lua/fidget.lua	2023-03-28 11:27:47.176862227 +0300
+++ fidget-modified.nvim/lua/fidget.lua	2023-03-28 11:28:01.364968991 +0300
@@ -69,6 +69,7 @@
 }

 local fidgets = {}
+local ignored_tasks = {}


 local function ignore_E523(callable)
@@ -428,6 +429,24 @@
     return
   end

+  -- if filter matches, ignore future notifications from this task
+  if val.kind == 'begin' and options.filter then
+    if options.filter(client_name, val.title) ~= true then
+      ignored_tasks[task] = {}
+      return
+    end
+  end
+
+  -- clean up ignore list
+  if val.kind == 'end' and ignored_tasks[task] ~= nil then
+    ignored_tasks[task] = nil
+    return
+  end
+
+  if ignored_tasks[task] then
+    return
+  end
+
   -- Create entry if missing
   if fidgets[client_key] == nil then
     fidgets[client_key] = new_fidget(client_key, client_name)
